File: src/test_suite.py

from abc import ABC, abstractmethod

class TestSuite(ABC):

    n: int

    def __init__(self, n: int) -> None:
        self.n = n

    @abstractmethod
    def generate_tests(self) -> list[int]:
        pass


class ListTestSuite(TestSuite):

    def generate_tests(self) -> list[int]:
        return list(range(self.n))

File: src/test-suite.cc
#include "test-suite.h"
#include <iostream>

TestSuite::TestSuite(uint32_t nodes)
    : n { nodes }
{
}

ListTestSuite::ListTestSuite(uint32_t nodes)
    : TestSuite { nodes }
{
}

std::vector<uint32_t>
ListTestSuite::generate_tests()
{
    std::vector<uint32_t> v(n);
    for (uint32_t i = 0; i < n; ++i)
        v[i] = i;
    return v;
}
